presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java

```diff
@@ -0,0 +1,295 @@
+                elseResult = new ConstantExpression(null, valueType);
+            }
+            else {
+                whenClauses = caseExpression.getArguments().subList(1, argumentsSize - 1);
+                elseResult = last;
+            }
+
+            if (caseOperand.isPresent() ?
+                    !canRewriteSimpleCaseExpression(whenClauses) :
+                    !canRewriteSearchedCaseExpression(whenClauses)) {
+                return null;
+            }
+
+            List<RowExpression> andExpressions = new ArrayList<>();
+            List<RowExpression> invertedOperands = new ArrayList<>();
+
+            for (RowExpression whenClause : whenClauses) {
+                RowExpression whenOperand = ((SpecialFormExpression) whenClause).getArguments().get(0);
+                if (caseOperand.isPresent()) {
+                    whenOperand = logicalRowExpressions.equalsCallExpression(caseOperand.get(), whenOperand);
+                }
+                RowExpression whenResult = ((SpecialFormExpression) whenClause).getArguments().get(1);
+                if (castExpression.isPresent()) {
+                    whenResult = replaceArguments(castExpression.get(), whenResult);
+                }
+
+                RowExpression comparisonExpression = comparisonExpressionGenerator.apply(whenResult);
+                andExpressions.add(and(comparisonExpression, whenOperand));
+                invertedOperands.add(logicalRowExpressions.notCallExpression(whenOperand));
+            }
+
+            RowExpression elseCondition = and(
+                    comparisonExpressionGenerator.apply(castExpression.map(x -> replaceArguments(x, elseResult)).orElse(elseResult)),
+                    and(invertedOperands));
+            andExpressions.add(elseCondition);
+
+            return or(andExpressions);
+        }
+
+        private Optional<RowExpression> getCaseOperand(SpecialFormExpression specialForm)
+        {
+            RowExpression expression = specialForm.getArguments().get(0);
+            boolean searchedCase = (expression instanceof ConstantExpression && expression.getType() == BOOLEAN &&
+                    ((ConstantExpression) expression).getValue() == Boolean.TRUE);
+            return searchedCase ? Optional.empty() : Optional.of(expression);
+        }
+
+        private boolean canRewriteSimpleCaseExpression(List<RowExpression> whenClauses)
+        {
+            List<RowExpression> whenOperands = whenClauses.stream()
+                    .map(x -> ((SpecialFormExpression) x).getArguments().get(0))
+                    .collect(Collectors.toList());
+            return validateRHSExpressions(whenOperands);
+        }
+
+        private boolean canRewriteSearchedCaseExpression(List<RowExpression> whenClauses)
+        {
+            RowExpression commonLHSExpression = null;
+            List<RowExpression> rightHandSideExpressions = new ArrayList<>();
+            for (RowExpression whenClause : whenClauses) {
```

Comment:
I find this a little tough to read (need to really work through the logic to understand what's going on)  I think it would be clearer to split up the three things going on in this loop even if it's less efficient.
e.g. something like
```

List<RowExpression> whenEqualityOperands = whenClauses.stream()
.map(whenClause -> (SpecialFormExpression) whenClause).getArguments().get(0))
.filter(logicalRowExpressions.isEqualExpression())
.collect(toImmutableList())

if(!whenEqualityOperands.size() == whenClauses.size()) {
   return false;
}

distinctLeftArgumentsCount = whenEqualityOperands.stream()
.map(whenOperand -> (CallExpression) whenOperand).getArguments().get(0))
.distinct()
.count()

if(!distinctLeftArgumentsCount == 1)
{
  return false;
}

List<RowExpression> rightArguments = whenEqualityOperands.stream()
.map(whenOperand -> ((CallExpression) whenOperand).getArguments().get(1))
.collect(toImmutableList())

return validateRHSExpressions(rightArguments);
```

Reply:
Broke the method into smaller ones.